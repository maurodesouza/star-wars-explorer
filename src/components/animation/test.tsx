import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Animation, AnimationProps } from '.';

jest.mock('react-lottie', () => ({
  __esModule: true,
  default: function Mock() {
    return <svg data-testid="lottie" />;
  },
}));

const dataMock: AnimationProps = {
  animation: 'loading',
  label: 'foo',
};

describe('<Animation />', () => {
  it('should render the animation styles correct', () => {
    const { container } = renderWithProviders(<Animation {...dataMock} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the animation component correct', () => {
    renderWithProviders(<Animation {...dataMock} />);

    const label = screen.getByText(dataMock.label!);
    const lottie = screen.getByTestId('lottie');

    expect(label).toBeInTheDocument();
    expect(lottie).toBeInTheDocument();
  });
});
