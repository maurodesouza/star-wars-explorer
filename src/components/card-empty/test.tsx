import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { EmptyCard } from '.';

const dataMock = {
  id: 'test',
};

describe('<EmptyCard />', () => {
  it('should render the empty card styles correct', () => {
    const { container } = renderWithProviders(<EmptyCard {...dataMock} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the empty card component correct', () => {
    renderWithProviders(<EmptyCard {...dataMock} />);

    const label = screen.getByText(`${dataMock.id}#`);

    expect(label).toBeInTheDocument();
  });
});
