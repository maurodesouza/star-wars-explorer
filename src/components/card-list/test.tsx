import { cleanup } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Entities } from 'types';
import { createArray } from 'utils';

import { CardList } from '.';

jest.mock('components', () => ({
  __esModule: true,
  Card: function Mock() {
    return <div data-testid="card" />;
  },
}));

const itemMock = {
  id: '1',
  title: 'C-3PO',
  image: 'image',
  entity: Entities.CHARACTERS,
};

describe('<CardList />', () => {
  describe('styles', () => {
    beforeAll(() => {
      jest.resetAllMocks();
    });

    it('should render the card list styles correct', () => {
      const { container } = renderWithProviders(<CardList />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  it('should render the items card correct', () => {
    const counts = [4, 8];

    counts.forEach(count => {
      const items = createArray(count).map((_, index) => ({
        ...itemMock,
        id: String(index),
      }));

      const render = renderWithProviders(<CardList items={items} />);
      const findedItems = render.getAllByTestId(/card/i);

      expect(findedItems.length).toBe(count);

      cleanup();
    });
  });

  it('should render the fills correct', () => {
    const minCountToFill = 4;
    const counts = [1, 3, 4, 6];

    counts.forEach(count => {
      const result = count - minCountToFill;
      const lackForFill = result <= 0 ? Math.abs(result) : 0;

      const items = createArray(count).map((_, index) => ({
        ...itemMock,
        id: String(index),
      }));

      const { container } = renderWithProviders(<CardList items={items} />);
      const findedItems = container.firstElementChild!.querySelectorAll(
        'div:not([data-testId]'
      );

      expect(findedItems.length).toBe(lackForFill);
      cleanup();
    });
  });
});
