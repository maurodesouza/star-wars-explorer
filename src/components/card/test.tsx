/* eslint-disable @typescript-eslint/no-var-requires */

import { act, fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Entities } from 'types';
import { Card } from '.';

const useRouter = jest.spyOn(require('next/router'), 'useRouter');
const push = jest.fn();

useRouter.mockImplementation(() => ({
  push,
}));

const mockData = {
  id: '1',
  label: 'C-3PO',
  image:
    'https://raw.githubusercontent.com/tbone849/star-wars-guide/master/build/assets/img/characters/2.jpg',
};

describe('<Card />', () => {
  it('should render the card correct', () => {
    renderWithProviders(<Card {...mockData} />);

    expect(
      screen.getByRole('img', { name: mockData.label })
    ).toBeInTheDocument();

    expect(
      screen.getByRole('heading', { name: mockData.label })
    ).toBeInTheDocument();
  });

  it("should render the no image found element when can't load image url", async () => {
    renderWithProviders(<Card {...mockData} />);

    const image = screen.getByRole('img', { name: mockData.label });

    act(() => {
      fireEvent.error(image);
    });

    const noImageFoundElement = screen.getByText(/no found/i);

    expect(noImageFoundElement).toBeInTheDocument();
  });

  it('should render the card styles correct', () => {
    const { container } = renderWithProviders(<Card {...mockData} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should navigate to entity page', () => {
    const { container } = renderWithProviders(<Card {...mockData} />);

    fireEvent.click(container.firstChild!);

    expect(push).toBeCalledTimes(1);
    expect(push).toBeCalledWith(`/${Entities.CHARACTERS}/${mockData.label}`);
  });
});
