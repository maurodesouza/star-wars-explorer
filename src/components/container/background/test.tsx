import { renderWithProviders } from 'utils/test/helpers';
import { Background } from '.';

describe('components/container', () => {
  describe('<Background />', () => {
    it('should render the background styles correct', () => {
      const { container } = renderWithProviders(<Background />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
