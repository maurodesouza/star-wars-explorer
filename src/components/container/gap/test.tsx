import { renderWithProviders } from 'utils/test/helpers';
import { Gap } from '.';

describe('components/container', () => {
  describe('<Gap />', () => {
    it('should render the gap styles correct', () => {
      const { container } = renderWithProviders(<Gap />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
