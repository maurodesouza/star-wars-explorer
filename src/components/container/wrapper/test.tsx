import { renderWithProviders } from 'utils/test/helpers';
import { Wrapper } from '.';

describe('components/container', () => {
  describe('<Wrapper />', () => {
    it('should render the wrapper styles correct', () => {
      const { container } = renderWithProviders(<Wrapper />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
