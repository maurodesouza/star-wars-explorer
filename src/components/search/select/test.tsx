import { act, fireEvent, screen } from '@testing-library/react';

import { renderWithProviders } from 'utils/test/helpers';

import { events } from 'app';
import { Entities } from 'types';

import { Select } from '.';

jest.useFakeTimers();

describe('<Select />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  it('should render the select styles correct', () => {
    const { container } = renderWithProviders(<Select />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should open dropdown select on input click and hide on blur', () => {
    const { container } = renderWithProviders(<Select />);

    const input = screen.getByLabelText(/open drop down select/i);

    const dropdown = container.children[0].children[2];

    expect(dropdown.getAttribute('aria-hidden')).toBe('true');

    act(() => {
      fireEvent.click(input);
    });

    expect(dropdown.getAttribute('aria-hidden')).toBe('false');

    act(() => {
      fireEvent.blur(input);
    });

    expect(dropdown.getAttribute('aria-hidden')).toBe('true');
  });

  it('should close dropdown after select some option', () => {
    const { container } = renderWithProviders(<Select />);

    const input = screen.getByLabelText(/open drop down select/i);

    const dropdown = container.children[0].children[2];

    act(() => {
      fireEvent.click(input);
    });

    act(() => {
      fireEvent.click(dropdown.children[0]);
    });

    expect(dropdown.getAttribute('aria-hidden')).toBe('true');
  });

  it('should send search make event on select some option', async () => {
    const searchMakeSpy = jest.spyOn(events.search, 'make');
    renderWithProviders(<Select />);

    const input = screen.getByLabelText(/open drop down select/i);

    act(() => {
      fireEvent.click(input);
    });

    const option = screen.getByText(Entities.PLANETS);

    act(() => {
      fireEvent.click(option);
    });

    expect(searchMakeSpy).toBeCalledTimes(1);
    expect(searchMakeSpy).toBeCalledWith({ entity: Entities.PLANETS });
  });
});
