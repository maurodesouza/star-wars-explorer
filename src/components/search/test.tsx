import { act, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithProviders } from 'utils/test/helpers';

import { events } from 'app';
import { sleep } from 'utils';

import { Search, SearchProps } from '.';

jest.mock('./select', () => ({
  __esModule: true,
  Select: function Mock() {
    return <div data-testid="search select" />;
  },
}));

const searchProps: Required<SearchProps> = {
  label: 'label',
  placeholder: 'placeholder',
};

describe('<Search />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the search correct', () => {
    renderWithProviders(<Search {...searchProps} />);

    const input = screen.getByPlaceholderText(searchProps.placeholder);
    const label = screen.getByLabelText(searchProps.label);
    const select = screen.getByTestId(/search select/i);

    expect(input).toBeInTheDocument();
    expect(select).toBeInTheDocument();
    expect(label).toBeInTheDocument();
  });

  it('should render the search styles correct', () => {
    const { container } = renderWithProviders(<Search />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render error', () => {
    const errorText = 'some error';

    renderWithProviders(<Search />);

    act(() => {
      events.search.error(errorText);
    });

    expect(screen.getByText(errorText)).toBeInTheDocument();
  });

  it('should send search event with corret values after type', async () => {
    const searchMakeEventFn = jest.spyOn(events.search, 'make');
    const text = 'some value';

    renderWithProviders(<Search {...searchProps} />);

    const input = screen.getByPlaceholderText(searchProps.placeholder);

    userEvent.type(input, text);
    await sleep(400);

    expect(searchMakeEventFn).toBeCalledWith({ search: text });
    expect(searchMakeEventFn).toBeCalledTimes(1);
  });

  it('should send search event with corret values after submit form with Enter', async () => {
    const searchMakeEventFn = jest.spyOn(events.search, 'make');

    renderWithProviders(<Search {...searchProps} />);

    const form = screen.getByRole('form');

    act(() => {
      fireEvent.submit(form);
    });

    expect(searchMakeEventFn).toBeCalledWith({ search: '' });
    expect(searchMakeEventFn).toBeCalledTimes(1);
  });
});
