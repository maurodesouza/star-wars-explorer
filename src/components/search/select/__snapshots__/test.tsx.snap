// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render the select styles correct 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  background-color: #FFE81F;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 0.8rem 0 0 0.8rem;
  position: relative;
  -webkit-transition: filter 0.2s;
  transition: filter 0.2s;
  color: #000;
  padding-inline: calc(1.2rem);
  width: 65%;
}

.c1 {
  width: 100%;
  border: 0;
  background: none;
  text-transform: capitalize;
  font-size: 1.6rem;
  font-weight: 500;
  color: inherit;
}

.c1::-webkit-input-placeholder {
  color: inherit;
}

.c1::-moz-placeholder {
  color: inherit;
}

.c1:-ms-input-placeholder {
  color: inherit;
}

.c1::placeholder {
  color: inherit;
}

.c3 {
  width: 1.2rem;
  height: 1.2rem;
  position: relative;
  top: 2px;
}

.c4 {
  z-index: 50;
  position: absolute;
  left: 0;
  right: 0;
  top: calc(100% + 1.2rem);
  width: 100%;
  background: #FFF;
  border-radius: 0.8rem;
  border-style: solid;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
  max-height: 20rem;
  opacity: 0;
  -webkit-transform: translateY(10px);
  -ms-transform: translateY(10px);
  transform: translateY(10px);
  -webkit-transition: 0.2s;
  transition: 0.2s;
  pointer-events: none;
}

.c4::-webkit-scrollbar-track {
  background: #FFF;
  border-radius: 0 0.8rem 0.8rem 0;
}

.c4::-webkit-scrollbar-thumb {
  border-radius: 0 calc(0.8rem / 2) calc(0.8rem / 2) 0;
}

.c5 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 100%;
  padding: calc(1.2rem / 2) 1.2rem;
  cursor: pointer;
  text-transform: capitalize;
  font-size: 1.2rem;
  background: #FFF;
  color: #000;
  border: none;
  text-align: left;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: 1;
  -webkit-box-orient: vertical;
}

.c5:hover {
  -webkit-filter: brightness(90%);
  filter: brightness(90%);
}

@media (min-width:450px) {
  .c0 {
    width: 50%;
  }
}

@media (min-width:768px) {
  .c0 {
    padding: 0 2.4rem;
    font-size: 2.4rem;
  }
}

@media (min-width:768px) {
  .c1 {
    font-size: 2.4rem;
  }
}

@media (min-width:768px) {
  .c3 {
    width: 1.6rem;
    height: 1.6rem;
    top: 4px;
  }
}

@media (min-width:768px) {
  .c5 {
    padding: 1.2rem 3.2rem;
    font-size: 1.6rem;
  }
}

<div
  class="c0"
>
  <input
    aria-label="Open drop down select"
    class="c1"
    placeholder="characters"
  />
  <svg
    aria-hidden="true"
    class="c2 c3"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 320 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M285.476 272.971 91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
      fill="currentColor"
    />
  </svg>
  <div
    aria-hidden="true"
    aria-label="drop down select"
    class="c4"
  >
    <button
      class="c5"
    >
      characters
    </button>
    <button
      class="c5"
    >
      films
    </button>
    <button
      class="c5"
    >
      species
    </button>
    <button
      class="c5"
    >
      starships
    </button>
    <button
      class="c5"
    >
      vehicles
    </button>
    <button
      class="c5"
    >
      planets
    </button>
  </div>
</div>
`;
